### PR TITLE
Add Triton Flash Attention 2 forward op

### PR DESCRIPTION
**Pull request by sgrigory**
Add an implementation of Triton Flash Attention 2 forward op supporting causal, block-diagonal causal, and padded block-diagonal causal masks. This op can be used in the prefill phase of LLM inference.

This implementation is based on a combination of the sources below. Block-diagonal mask support is new.

- https://github.com/openai/triton/blob/293b7fd592a1602f2305c1bd0bc978bbd97337d6/python/tutorials/06-fused-attention.py
- https://github.com/openai/triton/blob/293b7fd592a1602f2305c1bd0bc978bbd97337d6/python/triton/ops/flash_attention.py 
- https://github.com/Dao-AILab/flash-attention/blob/dd9a6fa45a9b90ff954d2b3f3f44241b9216190e/flash_attn/flash_attn_triton.py
- https://github.com/ROCmSoftwarePlatform/triton/blob/670ae8054da008424097989a5b6e3816aa601e07/python/perf-kernels/06-fused-attention-transV.py

Note this PR removes the backwards op, which was disabled anyway. Once we have a working version of backwards, we can add it separately.

# Tests

```
pytest tests/test_mem_eff_attention.py -vvv -k tritonflashattF

============== 455 passed, 704 skipped, 748 deselected in 19.90s ===============
```
https://gist.github.com/sgrigory/28c80629b61e06398fec514cc96e0508

# Benchmarks
For LLaMA shapes, with local changes to skip backward and only select these specific shapes
```
python xformers/benchmarks/benchmark_mem_eff_attention.py --omit-baselines 2>&1 | tee log.txt
```
```

[------------------ attention (attn_bias=<class 'NoneType'>) -----------------]
                                                |    ckF     |  tritonflashattF
1 threads: --------------------------------------------------------------------
      f16 1-2048-8-128, p=0.0, BiasT=NoneType   |     444.2  |         452.8   
      f16 1-8192-8-128, p=0.0, BiasT=NoneType   |    4426.4  |        3300.5   
      f16 4-2048-8-128, p=0.0, BiasT=NoneType   |    1164.6  |        1051.0   
      f16 4-8192-8-128, p=0.0, BiasT=NoneType   |   17686.3  |       11054.4   
      f16 8-2048-8-128, p=0.0, BiasT=NoneType   |    2313.8  |        1663.9   
      f16 8-8192-8-128, p=0.0, BiasT=NoneType   |   35602.5  |       22058.0   
      f16 16-2048-8-128, p=0.0, BiasT=NoneType  |    4596.5  |        3135.7   
      f16 16-8192-8-128, p=0.0, BiasT=NoneType  |   70172.1  |       43762.1   
      f16 32-2048-8-128, p=0.0, BiasT=NoneType  |    9195.3  |        6083.3   
      f16 32-8192-8-128, p=0.0, BiasT=NoneType  |  140831.0  |       86260.1   
      f16 64-2048-8-128, p=0.0, BiasT=NoneType  |   18120.0  |       11918.9   
      f16 64-8192-8-128, p=0.0, BiasT=NoneType  |  282569.4  |      173456.5   

Times are in microseconds (us).

[--- attention (attn_bias=<class 'xformers.ops.fmha.attn_bias.LowerTriangularMask'>) ----]
                                                           |    ckF     |  tritonflashattF
1 threads: -------------------------------------------------------------------------------
      f16 1-2048-8-128, p=0.0, BiasT=LowerTriangularMask   |     343.2  |         469.0   
      f16 1-8192-8-128, p=0.0, BiasT=LowerTriangularMask   |    2885.9  |        2342.6   
      f16 4-2048-8-128, p=0.0, BiasT=LowerTriangularMask   |     904.6  |         992.3   
      f16 4-8192-8-128, p=0.0, BiasT=LowerTriangularMask   |   10576.2  |        7546.5   
      f16 8-2048-8-128, p=0.0, BiasT=LowerTriangularMask   |    1768.0  |        1649.5   
      f16 8-8192-8-128, p=0.0, BiasT=LowerTriangularMask   |   20953.3  |       14452.8   
      f16 16-2048-8-128, p=0.0, BiasT=LowerTriangularMask  |    3490.1  |        3121.2   
      f16 16-8192-8-128, p=0.0, BiasT=LowerTriangularMask  |   41596.6  |       28055.4   
      f16 32-2048-8-128, p=0.0, BiasT=LowerTriangularMask  |    6919.4  |        6059.5   
      f16 32-8192-8-128, p=0.0, BiasT=LowerTriangularMask  |   82730.5  |       55212.8   
      f16 64-2048-8-128, p=0.0, BiasT=LowerTriangularMask  |   14051.8  |       12036.7   
      f16 64-8192-8-128, p=0.0, BiasT=LowerTriangularMask  |  166060.1  |      110122.3   

Times are in microseconds (us).
```

Note that this benchmark doesn't support multiquery and assumes that q/k/v are in "packed" layout. It also doesn't support block-diagonal masks supported by CK and Triton FA2 kernels. After the update from upstream https://github.com/ROCmSoftwarePlatform/xformers/pull/5 is merged these features should become available I think

### CODE REVIEW

**Comment by tenpercent** [comment]
I was able to run and pass the tests locally (via https://github.com/ROCmSoftwarePlatform/phantom_amd_llm_operators/tree/master/operators/xformers/triton)

For the benchmarks, without local changes I get a crash with the shape 

```
====== {'shape': (256, 4096, 16, 64), 'num_threads': 1, 'dropout_p': 0.0, 'attn_bias_cfg': (<class 'NoneType'>, False), 'dtype': torch.float16} ======                                                                                                                                                                           
 22%|██████████████████████████████████████████████████████████████                                                                                                                                                                                                                             | 25/114 [00:35<01:22,  1.07it/s]Memory access fault by GPU node-4 (Agent handle: 0x6e848a0) on address 0x7f62cd48f000. Reason: Unknown.
Aborted (core dumped)
```

Should we change something in the benchmarking code so that it doesn't crash?

**Review by tenpercent — CHANGES_REQUESTED** [review]
Following up, if you could run the benchmarking code without local changes, and whether you get a crash or it successfully runs all benchmarking cases

**Comment by sgrigory** [comment]
Thanks for catching this. It was actually CK kernel which was throwing illegal memory access on shape `(256, 4096, 16, 64)`. It seems the error happens when the product of the first two dimensions of the output is >= 2**20. Perhaps some index is overflowing or the launch grid become too large in that case? We can follow up with authors of CK. For now I added a check for this into `ops.fmha.ck.FwOp.not_supported_reasons`, so the benchmark skips this shape for this kernel.
If you'll be reproducing this error in tests, just disable computation of reference attention, otherwise it'll OOM

After this (and a change in how OOMs are caught) the benchmark runs without local changes: https://gist.github.com/sgrigory/c328a39a2a2e769eeb0638d28dcc0a11

**Comment by tenpercent** [comment]
Thanks! I was able to complete a benchmark on MI250X, here's the result for posterity
```
Times are in milliseconds (ms).

[------------------ attention (attn_bias=<class 'NoneType'>) ------------------]                                                                                                                                       
                                                 |    ckF     |  tritonflashattF
1 threads: ---------------------------------------------------------------------
      f16 384-197-1-88, p=0.0, BiasT=NoneType    |     259.6  |                 
      f16 384-197-1-80, p=0.0, BiasT=NoneType    |     253.9  |                 
      f16 384-197-1-64, p=0.0, BiasT=NoneType    |     132.1  |         279.6   
      f16 1024-197-1-88, p=0.0, BiasT=NoneType   |     621.7  |                 
      f16 1024-197-1-80, p=0.0, BiasT=NoneType   |     614.2  |                 
      f16 1024-197-1-64, p=0.0, BiasT=NoneType   |     302.4  |         479.1   
      f16 512-197-1-80, p=0.0, BiasT=NoneType    |     312.6  |                 
      f16 32-197-16-80, p=0.0, BiasT=NoneType    |     312.1  |                 
      f16 32-197-16-64, p=0.0, BiasT=NoneType    |     168.6  |         309.6   
      f16 32-197-16-128, p=0.0, BiasT=NoneType   |     367.9  |        1840.5   
      f16 256-197-1-88, p=0.0, BiasT=NoneType    |     165.0  |                 
      f16 16-197-16-88, p=0.0, BiasT=NoneType    |     164.8  |                 
      f16 16-197-16-64, p=0.0, BiasT=NoneType    |      92.5  |         240.6   
      f16 16-197-16-128, p=0.0, BiasT=NoneType   |     190.4  |        1008.6   
      f16 1024-82-8-64, p=0.0, BiasT=NoneType    |     762.1  |        1547.6   
      f16 150-256-16-64, p=0.0, BiasT=NoneType   |     787.7  |         716.8   
      f16 64-256-12-64, p=0.0, BiasT=NoneType    |     279.1  |         340.6   
      f16 1-4096-16-40, p=0.0, BiasT=NoneType    |     933.6  |                 
      f16 1-16384-16-40, p=0.0, BiasT=NoneType   |   13972.1  |                 
      f16 1-4096-16-80, p=0.0, BiasT=NoneType    |    1868.6  |                 
      f16 1-16384-16-80, p=0.0, BiasT=NoneType   |   30228.6  |                 
      f16 4-4096-16-40, p=0.0, BiasT=NoneType    |    3459.8  |                 
      f16 4-16384-16-40, p=0.0, BiasT=NoneType   |   55479.6  |                 
      f16 4-4096-16-80, p=0.0, BiasT=NoneType    |    7417.4  |                 
      f16 4-16384-16-80, p=0.0, BiasT=NoneType   |  119900.4  |                 
      f16 256-4096-16-64, p=0.0, BiasT=NoneType  |            |      175979.6   
      f16 8-2048-20-128, p=0.0, BiasT=NoneType   |    5222.2  |        3544.3   
      f16 1-2048-4-128, p=0.0, BiasT=NoneType    |     210.3  |         323.0   
      f16 1-2048-8-128, p=0.0, BiasT=NoneType    |     421.1  |         394.9   
      f16 1-4096-4-128, p=0.0, BiasT=NoneType    |     805.1  |         617.7   
      f16 1-4096-8-128, p=0.0, BiasT=NoneType    |    1223.8  |        1103.0   
      f16 1-8192-4-128, p=0.0, BiasT=NoneType    |    2407.2  |        2014.0   
      f16 1-8192-8-128, p=0.0, BiasT=NoneType    |    4077.3  |        2956.0   
      f16 2-2048-4-128, p=0.0, BiasT=NoneType    |     421.0  |         401.2   
      f16 2-2048-8-128, p=0.0, BiasT=NoneType    |     633.9  |         664.3   
      f16 2-4096-4-128, p=0.0, BiasT=NoneType    |    1219.6  |        1126.6   
      f16 2-4096-8-128, p=0.0, BiasT=NoneType    |    2050.0  |        1627.1   
      f16 2-8192-4-128, p=0.0, BiasT=NoneType    |    4050.4  |        3018.0   
      f16 2-8192-8-128, p=0.0, BiasT=NoneType    |    8114.9  |        5025.6   
      f16 16-128-16-16, p=0.0, BiasT=NoneType    |      60.5  |                 
      f16 16-128-16-32, p=0.0, BiasT=NoneType    |      51.8  |         163.7   
      f16 16-128-16-64, p=0.0, BiasT=NoneType    |      51.7  |         135.1   
      f16 16-128-16-128, p=0.0, BiasT=NoneType   |      86.7  |         216.6   
      f16 16-128-16-160, p=0.0, BiasT=NoneType   |     137.0  |                 
      f16 16-128-16-256, p=0.0, BiasT=NoneType   |     169.5  |                 
      f16 16-512-16-16, p=0.0, BiasT=NoneType    |     158.9  |                 
      f16 16-512-16-32, p=0.0, BiasT=NoneType    |     164.4  |         276.0   
      f16 16-512-16-64, p=0.0, BiasT=NoneType    |     288.2  |         239.3   
      f16 16-512-16-128, p=0.0, BiasT=NoneType   |     635.2  |         528.4   
      f16 16-512-16-160, p=0.0, BiasT=NoneType   |    1296.4  |                 
      f16 16-512-16-256, p=0.0, BiasT=NoneType   |    1655.9  |                 
      f16 16-1024-16-16, p=0.0, BiasT=NoneType   |     547.0  |                 
      f16 16-1024-16-32, p=0.0, BiasT=NoneType   |     562.8  |         604.6   
      f16 16-1024-16-64, p=0.0, BiasT=NoneType   |     945.3  |         824.7   
      f16 16-1024-16-128, p=0.0, BiasT=NoneType  |    2220.7  |        1849.2   
      f16 16-1024-16-160, p=0.0, BiasT=NoneType  |    4813.5  |                 
      f16 16-1024-16-256, p=0.0, BiasT=NoneType  |    6105.1  |                 
      f16 384-197-1-88, p=0.3, BiasT=NoneType    |     464.0  |                 
      f16 384-197-1-64, p=0.3, BiasT=NoneType    |     169.1  |                 
      f16 1024-197-1-64, p=0.3, BiasT=NoneType   |     389.4  |                 
      f16 32-197-16-80, p=0.3, BiasT=NoneType    |     574.1  |                 
      b16 256-197-1-88, p=0.0, BiasT=NoneType    |     164.0  |                 
      f16 16-197-16-88, p=0.3, BiasT=NoneType    |     300.5  |                 
      f16 150-256-16-64, p=0.3, BiasT=NoneType   |     959.2  |                 
      f16 1-16384-16-40, p=0.3, BiasT=NoneType   |   18756.8  |                 
      b16 8-2048-20-128, p=0.0, BiasT=NoneType   |    5218.1  |        4316.2   
      b16 1-4096-4-128, p=0.0, BiasT=NoneType    |     805.8  |         752.3   
      b16 1-4096-8-128, p=0.0, BiasT=NoneType    |    1222.8  |        1364.8   
      f16 2-2048-8-128, p=0.3, BiasT=NoneType    |    1289.2  |                 
      b16 2-4096-4-128, p=0.0, BiasT=NoneType    |    1219.0  |        1366.8   
      b16 2-8192-8-128, p=0.0, BiasT=NoneType    |    8130.0  |        6131.0   
      b16 16-128-16-128, p=0.0, BiasT=NoneType   |      86.2  |         229.6   
      f16 16-128-16-256, p=0.3, BiasT=NoneType   |     266.8  |                 
      b16 16-512-16-160, p=0.0, BiasT=NoneType   |    1294.1  |                 
      b16 16-1024-16-16, p=0.0, BiasT=NoneType   |     516.8  |                 
      f16 16-1024-16-32, p=0.3, BiasT=NoneType   |     895.2  |                 
      b16 16-1024-16-64, p=0.0, BiasT=NoneType   |     955.6  |        1150.2   
      b16 16-1024-16-128, p=0.0, BiasT=NoneType  |    2212.7  |        2120.0   
      b16 16-1024-16-256, p=0.0, BiasT=NoneType  |    6076.6  |                 

Times are in microseconds (us).

[--- attention (attn_bias=<class 'xformers.ops.fmha.attn_bias.LowerTriangularMask'>) ---]
                                                            |   ckF    |  tritonflashattF
1 threads: ------------------------------------------------------------------------------
      f16 384-197-1-80, p=0.0, BiasT=LowerTriangularMask    |   264.5  |                 
      f16 32-197-16-128, p=0.0, BiasT=LowerTriangularMask   |   387.6  |       1870.0    
      f16 16-197-16-128, p=0.0, BiasT=LowerTriangularMask   |   196.2  |       1153.6    
      f16 1-4096-16-40, p=0.0, BiasT=LowerTriangularMask    |   698.9  |                 
      f16 4-4096-16-40, p=0.0, BiasT=LowerTriangularMask    |  2363.5  |                 
      f16 2-8192-4-128, p=0.0, BiasT=LowerTriangularMask    |  2656.8  |       2102.4    
      f16 16-128-16-64, p=0.0, BiasT=LowerTriangularMask    |    52.4  |        177.8    
      f16 16-512-16-128, p=0.0, BiasT=LowerTriangularMask   |   653.3  |        597.0    
      f16 16-1024-16-160, p=0.0, BiasT=LowerTriangularMask  |  4879.1  |                 

Times are in microseconds (us).

[----- attention (attn_bias=<class 'torch.Tensor'>) ----]
                                              |    ckF   
1 threads: ----------------------------------------------
      f16 1-16384-16-80, p=0.0, BiasT=Tensor  |   70797.5
      f16 4-4096-16-80, p=0.0, BiasT=Tensor   |   16342.9
      f16 4-16384-16-80, p=0.0, BiasT=Tensor  |  282922.0
      f16 1-2048-8-128, p=0.0, BiasT=Tensor   |     890.8
      f16 2-4096-8-128, p=0.0, BiasT=Tensor   |    4430.1
      f16 16-512-16-16, p=0.0, BiasT=Tensor   |     485.3
      f16 16-512-16-32, p=0.0, BiasT=Tensor   |     496.3
      f16 16-512-16-64, p=0.0, BiasT=Tensor   |     660.4

Times are in microseconds (us).
```

File: xformers/ops/fmha/ck.py [hunk 1]
```diff
@@ -86,6 +86,20 @@ def _check_bias_alignment(
                 "you should call `.contiguous()` on the bias"
             )
 
+def _check_large_shapes(reasons: List[str], inp: Inputs) -> None:
```
Comment (tenpercent) [inline xformers/ops/fmha/ck.py:89]: cc @qianfengz

**Review by tenpercent — APPROVED** [review]
I think it's good to go; the tests are passing, and the benchmark is functional. We will need to follow up to figure what's going in CK with large input shapes, the only change in `ck.py` is work around an illegal memory access for these shapes.

**Review by qianfengz — APPROVED** [review]

**Pull request merged**